File: internal/usage/aggregator.go
package usage

import (
	"sort"
	"time"
)

// Aggregator accumulates parsed records into a WorkspaceUsage snapshot.
// One Aggregator corresponds to one transcript file (one session). When
// the session changes, callers should Reset it.
type Aggregator struct {
	sessionID    string
	sessionStart *time.Time
	lastActivity *time.Time

	cumulative    Tokens
	cumulativeMCP MCPProxyMetrics
	byModel       map[string]*ModelTotals
	currentTokens Tokens
	currentMCP    MCPProxyMetrics
	currentModel  string
	turns         int64

	parseErrors int64
	requests    map[string]*assistantRequestState
}

type assistantRequestState struct {
	model    string
	tokens   Tokens
	usesMCP  bool
	mcpProxy MCPProxyMetrics
}

type ingestResult struct {
	UsageObserved bool
	TokensDelta   Tokens
	MCPDelta      MCPProxyMetrics
	TurnDelta     int64
}

// NewAggregator returns an empty aggregator.
func NewAggregator() *Aggregator {
	return &Aggregator{
		byModel:  map[string]*ModelTotals{},
		requests: map[string]*assistantRequestState{},
	}
}

// Reset clears accumulated state (used when the active session file changes).
func (a *Aggregator) Reset() {
	*a = Aggregator{
		byModel:  map[string]*ModelTotals{},
		requests: map[string]*assistantRequestState{},
	}
}

// SessionID returns the most recent session id observed, or "".
func (a *Aggregator) SessionID() string { return a.sessionID }

// ParseErrors returns the count of malformed lines encountered.
func (a *Aggregator) ParseErrors() int64 { return a.parseErrors }

// Turns returns the number of usage-bearing turns ingested.
func (a *Aggregator) Turns() int64 { return a.turns }

// Ingest folds a parsed record into the running totals and returns the
// effective delta after request-level de-duplication.
func (a *Aggregator) Ingest(r parsedRecord) ingestResult {
	result := ingestResult{UsageObserved: r.HasUsage}
	if r.SessionID != "" {
		a.sessionID = r.SessionID
	}
	if !r.Timestamp.IsZero() {
		ts := r.Timestamp
		if a.sessionStart == nil {
			tsCopy := ts
			a.sessionStart = &tsCopy
		}
		tsCopy := ts
		a.lastActivity = &tsCopy
	}
	if !r.HasUsage {
		if r.MCPProxy != (MCPProxyMetrics{}) {
			a.cumulativeMCP = a.cumulativeMCP.Add(r.MCPProxy)
			a.currentMCP = r.MCPProxy
			result.MCPDelta = r.MCPProxy
		}
		return result
	}

	a.currentTokens = r.Tokens
	a.currentModel = r.Model

	key := r.requestKey()
	if key == "" {
		result.TokensDelta = r.Tokens
		result.MCPDelta = r.MCPProxy
		result.TurnDelta = 1
		a.applyUsageDelta("", Tokens{}, r.Model, r.Tokens, result.TurnDelta)
		if result.MCPDelta != (MCPProxyMetrics{}) {
			a.cumulativeMCP = a.cumulativeMCP.Add(result.MCPDelta)
			a.currentMCP = result.MCPDelta
		}
		return result
	}

	state, ok := a.requests[key]
	if !ok {
		state = &assistantRequestState{}
		a.requests[key] = state
		result.TurnDelta = 1
	}

	prevModel := state.model
	prevTokens := state.tokens
	prevMCP := state.mcpProxy
	if r.MCPProxy.ToolUseTurns > 0 {
		state.usesMCP = true
	}
	state.model = r.Model
	state.tokens = r.Tokens
	if state.usesMCP {
		total := state.tokens.Total()
		state.mcpProxy = MCPProxyMetrics{
			Total:         total,
			ToolUseTokens: total,
			ToolUseTurns:  1,
		}
	} else {
		state.mcpProxy = MCPProxyMetrics{}
	}

	result.TokensDelta = state.tokens.Sub(prevTokens)
	a.applyUsageDelta(prevModel, prevTokens, state.model, state.tokens, result.TurnDelta)
	result.MCPDelta = state.mcpProxy.Sub(prevMCP)
	if result.MCPDelta != (MCPProxyMetrics{}) {
		a.cumulativeMCP = a.cumulativeMCP.Add(result.MCPDelta)
		a.currentMCP = state.mcpProxy
	}
	return result
}

// IngestLine parses and ingests a single jsonl line. Malformed lines
// bump ParseErrors and return false. Returns true on usage-bearing lines.
func (a *Aggregator) IngestLine(line []byte) bool {
	rec, err := parseLine(line)
	if err != nil {
		a.parseErrors++
		return false
	}
	return a.Ingest(rec).UsageObserved
}

func (a *Aggregator) applyUsageDelta(prevModel string, prevTokens Tokens, nextModel string, nextTokens Tokens, turnDelta int64) {
	a.cumulative = a.cumulative.Add(nextTokens.Sub(prevTokens))
	a.turns += turnDelta

	switch {
	case prevModel == "" || prevModel == nextModel:
		mt := a.ensureModel(nextModel)
		mt.Totals = mt.Totals.Add(nextTokens.Sub(prevTokens))
		mt.Turns += turnDelta
	default:
		prev := a.ensureModel(prevModel)
		prev.Totals = prev.Totals.Sub(prevTokens)
		prev.Turns--
		if prev.Turns == 0 && prev.Totals == (Tokens{}) {
			delete(a.byModel, prevModel)
		}

		next := a.ensureModel(nextModel)
		next.Totals = next.Totals.Add(nextTokens)
		next.Turns++
	}
}

func (a *Aggregator) ensureModel(model string) *ModelTotals {
	mt, ok := a.byModel[model]
	if ok {
		return mt
	}
	mt = &ModelTotals{Model: model}
	a.byModel[model] = mt
	return mt
}

// Snapshot materializes the current state into a WorkspaceUsage value.
// ByModel is sorted by model name for deterministic output. Available is
// set true when at least one usage-bearing record or any sessionId has
// been observed.
func (a *Aggregator) Snapshot(workspace, transcriptPath string) WorkspaceUsage {
	models := make([]ModelTotals, 0, len(a.byModel))
	for _, m := range a.byModel {
		models = append(models, *m)
	}
	sort.Slice(models, func(i, j int) bool { return models[i].Model < models[j].Model })
	return WorkspaceUsage{
		Workspace:        workspace,
		TranscriptPath:   transcriptPath,
		SessionID:        a.sessionID,
		SessionStart:     a.sessionStart,
		LastActivity:     a.lastActivity,
		CumulativeTotals: a.cumulative,
		CumulativeMCP:    a.cumulativeMCP,
		ByModel:          models,
		CurrentContext:   a.currentTokens,
		CurrentMCP:       a.currentMCP,
		CurrentModel:     a.currentModel,
		Turns:            a.turns,
		Available:        a.turns > 0 || a.sessionID != "",
	}
}

File: crates/ax-usage/src/aggregator.rs
//! Incremental aggregation over transcript records.
//!
//! One [`Aggregator`] corresponds to one transcript file (one session).
//! [`Aggregator::ingest`] takes a [`ParsedRecord`] and folds it into the
//! running totals, deduping assistant requests by their `requestKey()` so
//! revision edits to the same turn update — not double-count — the
//! cumulative numbers.

use std::collections::BTreeMap;

use chrono::{DateTime, Utc};

use ax_proto::usage::{MCPProxyMetrics, ModelTotals, Tokens};

use crate::parse::{parse_line, ParseError, ParsedRecord};

/// Snapshot of the aggregator's state. Kept inside ax-usage for now
/// because the daemon wire representation for live per-workspace
/// usage hasn't been finalised.
#[derive(Debug, Clone, Default)]
pub struct UsageSnapshot {
    pub session_id: String,
    pub transcript_path: String,
    pub session_start: Option<DateTime<Utc>>,
    pub last_activity: Option<DateTime<Utc>>,
    pub cumulative_totals: Tokens,
    pub cumulative_mcp: MCPProxyMetrics,
    pub by_model: Vec<ModelTotals>,
    pub current_context: Tokens,
    pub current_mcp: MCPProxyMetrics,
    pub current_model: String,
    pub turns: i64,
    pub available: bool,
}

/// Effect of a single ingestion — deltas the history bucketing pass uses
/// to attribute this record to its time window.
#[derive(Debug, Clone, Copy, Default)]
pub struct IngestResult {
    pub usage_observed: bool,
    pub tokens_delta: Tokens,
    pub mcp_delta: MCPProxyMetrics,
    pub turn_delta: i64,
}

#[derive(Debug, Default)]
pub struct Aggregator {
    session_id: String,
    session_start: Option<DateTime<Utc>>,
    last_activity: Option<DateTime<Utc>>,

    cumulative: Tokens,
    cumulative_mcp: MCPProxyMetrics,
    by_model: BTreeMap<String, ModelTotals>,
    current_tokens: Tokens,
    current_mcp: MCPProxyMetrics,
    current_model: String,
    turns: i64,

    parse_errors: i64,
    requests: BTreeMap<String, AssistantRequestState>,
}

#[derive(Debug, Default, Clone)]
struct AssistantRequestState {
    model: String,
    tokens: Tokens,
    uses_mcp: bool,
    mcp_proxy: MCPProxyMetrics,
}

impl Aggregator {
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    /// Reset all accumulated state. Used when the active session file
    /// rotates mid-scan.
    pub fn reset(&mut self) {
        *self = Self::default();
    }

    #[must_use]
    pub fn session_id(&self) -> &str {
        &self.session_id
    }

    #[must_use]
    pub fn parse_errors(&self) -> i64 {
        self.parse_errors
    }

    #[must_use]
    pub fn turns(&self) -> i64 {
        self.turns
    }

    /// Parse `line` and ingest the result. Malformed lines bump
    /// [`Self::parse_errors`] and return `false`; otherwise returns the
    /// `usage_observed` bit of the effect.
    pub fn ingest_line(&mut self, line: &[u8]) -> bool {
        if let Ok(rec) = parse_line(line) {
            self.ingest(&rec).usage_observed
        } else {
            self.parse_errors += 1;
            false
        }
    }

    /// Fold a parsed record into the running totals.
    pub fn ingest(&mut self, rec: &ParsedRecord) -> IngestResult {
        let mut result = IngestResult {
            usage_observed: rec.has_usage,
            ..IngestResult::default()
        };
        if !rec.session_id.is_empty() {
            self.session_id.clone_from(&rec.session_id);
        }
        if let Some(ts) = rec.timestamp {
            if self.session_start.is_none() {
                self.session_start = Some(ts);
            }
            self.last_activity = Some(ts);
        }
        if !rec.has_usage {
            if rec.mcp_proxy != MCPProxyMetrics::default() {
                self.cumulative_mcp = add_mcp(self.cumulative_mcp, rec.mcp_proxy);
                self.current_mcp = rec.mcp_proxy;
                result.mcp_delta = rec.mcp_proxy;
            }
            return result;
        }

        self.current_tokens = rec.tokens;
        self.current_model.clone_from(&rec.model);

        let key = rec.request_key();
        if key.is_empty() {
            // No dedup key — treat as a fresh turn.
            result.tokens_delta = rec.tokens;
            result.mcp_delta = rec.mcp_proxy;
            result.turn_delta = 1;
            self.apply_usage_delta("", Tokens::default(), &rec.model, rec.tokens, 1);
            if result.mcp_delta != MCPProxyMetrics::default() {
                self.cumulative_mcp = add_mcp(self.cumulative_mcp, result.mcp_delta);
                self.current_mcp = result.mcp_delta;
            }
            return result;
        }

        // First occurrence → turn; subsequent updates to the same key
        // replace the previous snapshot so edits don't double-count.
        let (prev_model, prev_tokens, prev_mcp, new_mcp) = {
            let first_time = !self.requests.contains_key(&key);
            if first_time {
                result.turn_delta = 1;
            }
            let state = self.requests.entry(key).or_default();
            let prev_model = state.model.clone();
            let prev_tokens = state.tokens;
            let prev_mcp = state.mcp_proxy;
            if rec.mcp_proxy.tool_use_turns > 0 {
                state.uses_mcp = true;
            }
            state.model.clone_from(&rec.model);
            state.tokens = rec.tokens;
            state.mcp_proxy = if state.uses_mcp {
                let total = state.tokens.total();
                MCPProxyMetrics {
                    total,
                    tool_use_tokens: total,
                    tool_use_turns: 1,
                    ..MCPProxyMetrics::default()
                }
            } else {
                MCPProxyMetrics::default()
            };
            (prev_model, prev_tokens, prev_mcp, state.mcp_proxy)
        };

        result.tokens_delta = rec.tokens - prev_tokens;
        self.apply_usage_delta(
            &prev_model,
            prev_tokens,
            &rec.model,
            rec.tokens,
            result.turn_delta,
        );
        result.mcp_delta = sub_mcp(new_mcp, prev_mcp);
        if result.mcp_delta != MCPProxyMetrics::default() {
            self.cumulative_mcp = add_mcp(self.cumulative_mcp, result.mcp_delta);
            self.current_mcp = new_mcp;
        }
        result
    }

    /// Materialise the current state into a snapshot. `workspace` and
    /// `transcript_path` are attached unchanged so callers can record
    /// where the data came from.
    #[must_use]
    pub fn snapshot(&self, workspace: &str, transcript_path: &str) -> UsageSnapshot {
        let _ = workspace; // reserved for the eventual public shape
        let mut models: Vec<ModelTotals> = self.by_model.values().cloned().collect();
        models.sort_by(|a, b| a.model.cmp(&b.model));
        UsageSnapshot {
            session_id: self.session_id.clone(),
            transcript_path: transcript_path.to_owned(),
            session_start: self.session_start,
            last_activity: self.last_activity,
            cumulative_totals: self.cumulative,
            cumulative_mcp: self.cumulative_mcp,
            by_model: models,
            current_context: self.current_tokens,
            current_mcp: self.current_mcp,
            current_model: self.current_model.clone(),
            turns: self.turns,
            available: self.turns > 0 || !self.session_id.is_empty(),
        }
    }

    // ---------- internals ----------

    fn apply_usage_delta(
        &mut self,
        prev_model: &str,
        prev_tokens: Tokens,
        next_model: &str,
        next_tokens: Tokens,
        turn_delta: i64,
    ) {
        self.cumulative = self.cumulative + (next_tokens - prev_tokens);
        self.turns += turn_delta;
        if prev_model.is_empty() || prev_model == next_model {
            let mt = self.ensure_model(next_model);
            mt.totals = mt.totals + (next_tokens - prev_tokens);
            mt.turns += turn_delta;
            return;
        }
        // Model change: remove prev-model share, add next-model share.
        let prev = self.ensure_model(prev_model);
        prev.totals = prev.totals - prev_tokens;
        prev.turns -= 1;
        let should_drop = prev.turns == 0 && prev.totals == Tokens::default();
        if should_drop {
            self.by_model.remove(prev_model);
        }
        let next = self.ensure_model(next_model);
        next.totals = next.totals + next_tokens;
        next.turns += 1;
    }

    fn ensure_model(&mut self, model: &str) -> &mut ModelTotals {
        self.by_model
            .entry(model.to_owned())
            .or_insert_with(|| ModelTotals {
                model: model.to_owned(),
                turns: 0,
                totals: Tokens::default(),
            })
    }
}

fn add_mcp(a: MCPProxyMetrics, b: MCPProxyMetrics) -> MCPProxyMetrics {
    MCPProxyMetrics {
        total: a.total + b.total,
        prompt_tokens: a.prompt_tokens + b.prompt_tokens,
        prompt_signals: a.prompt_signals + b.prompt_signals,
        tool_use_tokens: a.tool_use_tokens + b.tool_use_tokens,
        tool_use_turns: a.tool_use_turns + b.tool_use_turns,
    }
}

fn sub_mcp(a: MCPProxyMetrics, b: MCPProxyMetrics) -> MCPProxyMetrics {
    MCPProxyMetrics {
        total: a.total - b.total,
        prompt_tokens: a.prompt_tokens - b.prompt_tokens,
        prompt_signals: a.prompt_signals - b.prompt_signals,
        tool_use_tokens: a.tool_use_tokens - b.tool_use_tokens,
        tool_use_turns: a.tool_use_turns - b.tool_use_turns,
    }
}

/// Compatibility shim so callers can feed an aggregator from raw JSONL
/// lines without pulling in [`parse_line`] directly.
pub fn ingest_line<'a>(
    agg: &'a mut Aggregator,
    line: &[u8],
) -> Result<&'a mut Aggregator, ParseError> {
    let rec = parse_line(line)?;
    agg.ingest(&rec);
    Ok(agg)
}
